vllm/distributed/device_communicators/pynccl.py

```diff
@@ -146,17 +148,38 @@ def all_gather(self,
             f"but the input tensor is on {input_tensor.device}")
         if stream is None:
             stream = current_stream()
-        self.nccl.ncclAllGather(
-            buffer_type(input_tensor.data_ptr()),
-            buffer_type(output_tensor.data_ptr()), input_tensor.numel(),
-            ncclDataTypeEnum.from_torch(input_tensor.dtype), self.comm,
-            cudaStream_t(stream.cuda_stream))
+        if sizes is not None:
+            assert output_tensor.shape[0] == sum(sizes)
+            numel_base = int(np.prod(output_tensor.shape[1:]))
+            split_offset = 0
+            self.nccl.ncclGroupStart()
```

Comment:
This looks interesting -- How do ncclGroupStart() and ncclGroupEnd() work?

Reply:
These functions allow you batch multiple NCCL calls into one single launch.
You can read more about it here: https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/groups.html